library/pk_wrap.c

```diff
@@ -1095,13 +1095,92 @@ static int eckey_sign_rs_wrap(void *ctx, mbedtls_md_type_t md_alg,
 }
 #endif /* MBEDTLS_ECDSA_C && MBEDTLS_ECP_RESTARTABLE */
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+/*
+ * Alternative function used to verify that the EC private/public key pair
+ * is valid using PSA functions instead of ECP ones.
+ * The flow is:
+ * - import the private key "prv" to PSA and export its public part
+ * - write the raw content of public key "pub" to a local buffer
+ * - compare the two buffers
+ */
+static int eckey_check_pair_psa(const void *pub, const void *prv)
+{
+    psa_status_t status;
+    psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+    mbedtls_ecp_keypair *prv_ctx = (mbedtls_ecp_keypair *) prv;
+    mbedtls_ecp_keypair *pub_ctx = (mbedtls_ecp_keypair *) pub;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    uint8_t prv_key_buf[MBEDTLS_PSA_MAX_EC_KEY_PAIR_LENGTH];
+    size_t prv_key_len;
+    uint8_t pub_key_buf[MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH];
+    size_t pub_key_len;
+    mbedtls_svc_key_id_t key_id = MBEDTLS_SVC_KEY_ID_INIT;
+    size_t curve_bits;
+    psa_ecc_family_t curve =
+        mbedtls_ecc_group_to_psa(prv_ctx->grp.id, &curve_bits);
+    size_t curve_bytes = PSA_BITS_TO_BYTES(curve_bits);
+
+    psa_set_key_type(&key_attr, PSA_KEY_TYPE_ECC_KEY_PAIR(curve));
+    psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_EXPORT);
+
+    ret = mbedtls_mpi_write_binary(&prv_ctx->d, prv_key_buf, curve_bytes);
+    if (ret != 0) {
+        return ret;
+    }
+
+    status = psa_import_key(&key_attr, prv_key_buf, curve_bytes, &key_id);
+    if (status != PSA_SUCCESS) {
+        ret = PSA_PK_TO_MBEDTLS_ERR(status);
+        return ret;
+    }
+
+    mbedtls_platform_zeroize(prv_key_buf, sizeof(prv_key_buf));
+
+    status = psa_export_public_key(key_id, prv_key_buf, sizeof(prv_key_buf),
```

Comment:
Right, indeed the failure found by the CI was exactly about this

Reply:
Ah, I wondered if it was related, but didn't investigate, thanks for the info!

Reply:
> the serialization format for an ECC "key pair" in PSA is actually only the private key, which is shorter than the public key

This is not guaranteed, although it is the case for all the curves I'm aware of. For Montgomery and Edwards curves, they have the same length.

Reply:
We have a number of cases in PK where we want a buffer that can hold either a private or a public ECC key. Do you think we should add a new macro for that?

I'm inclined to think that if something is true for all curvtes types now, it's likely to remain true for the foreseeable future, as new curves types seem unlikely to appear (these days innovation is rather focused on PQC schemes). OTOH, if it ever changes, of course it's better to have a single place to change (the definition of the new macro) rather than have to go over all uses of `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE` and decide if they need changing or not.

Reply:
Even if the risk of future bugs is low, having a distinct macro would be useful documentation for humans (and perhaps static analyzers): it's not obvious that `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE` is large enough for a private key. However, this only concerns a small part of the code (the part that's about ECC keys in general, excluding parts that know whether they're dealing with a key pair or a public key as well as parts that deal with all key types), so maybe it isn't worth it.